inc/sainat.h

```diff
@@ -0,0 +1,653 @@
+ *    a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR
+ *    CONDITIONS OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT
+ *    LIMITATION ANY IMPLIED WARRANTIES OR CONDITIONS OF TITLE, FITNESS
+ *    FOR A PARTICULAR PURPOSE, MERCHANTABILITY OR NON-INFRINGEMENT.
+ *
+ *    See the Apache Version 2.0 License for specific language governing
+ *    permissions and limitations under the License.
+ *
+ *    Microsoft would like to thank the following companies for their review and
+ *    assistance with these files: Intel Corporation, Mellanox Technologies Ltd,
+ *    Dell Products, L.P., Facebook, Inc., Marvell International Ltd.
+ *
+ * @file    sainat.h
+ *
+ * @brief   This module defines SAI NAT (Network Address Translation) spec
+ */
+
+#if !defined (__SAINAT_H_)
+#define __SAINAT_H_
+
+#include <saitypes.h>
+
+/**
+ * @defgroup SAINAT SAI - Network Address Translation (NAT) specific API definitions
+ *
+ * @{
+ */
+
+/**
+ * @brief NAT Type
+ */
+typedef enum _sai_nat_type_t
+{
+    /** No NAT */
+    SAI_NAT_TYPE_NONE,
+
+    /** Source NAT */
+    SAI_NAT_TYPE_SOURCE_NAT,
+
+    /** Destination NAT */
+    SAI_NAT_TYPE_DESTINATION_NAT,
+
+    /** Double NAT */
+    SAI_NAT_TYPE_DOUBLE_NAT,
+
+} sai_nat_type_t;
+
+/**
+ * @brief NAT Entry Attributes for Match
+ */
+typedef enum _sai_nat_entry_attr_t
+{
+
+    /**
+     * @brief Start of Attributes
+     */
+    SAI_NAT_ENTRY_ATTR_START,
+
+    /**
+     * @brief NAT Type defined in sai_nat_type_t
+     * @type sai_nat_type_t
+     * @flags CREATE_AND_SET
+     * @default SAI_NAT_TYPE_NONE
+     */
+    SAI_NAT_ENTRY_ATTR_NAT_TYPE  = SAI_NAT_ENTRY_ATTR_START,
+
+    /**
+     * @brief NAT Inside Zone ID
+     *
+     * @type sai_uint8_t
+     * @flags CREATE_AND_SET
+     * @default 0
+     */
+    SAI_NAT_ENTRY_ATTR_FROM_ZONE_ID,
+
+    /**
+     * @brief NAT Outside Zone ID
+     *
+     * @type sai_uint8_t
+     * @flags CREATE_AND_SET
+     * @default 0
+     */
+    SAI_NAT_ENTRY_ATTR_TO_ZONE_ID,
+
+    /**
+     * @brief Replace source IPv4 address in packet.
+     * NAT actions will be
+     *    (source/destination/both is identified by type of NAT)
+     *    - replace IP address
+     *    - replace layer 4 source port
+     *    - replace layer 4 destination port
+     *
+     * @type sai_ip4_t
+     * @flags CREATE_AND_SET
+     * @default 0.0.0.0
+     * @validonly SAI_NAT_ENTRY_ATTR_NAT_TYPE == SAI_NAT_TYPE_SOURCE_NAT or SAI_NAT_ENTRY_ATTR_NAT_TYPE == SAI_NAT_TYPE_DOUBLE_NAT
+     */
+    SAI_NAT_ENTRY_ATTR_SRC_IP,
+
+    /**
+     * @brief Mask for source IPv4 address in packet.
+     *
+     * @type sai_ip4_t
+     * @flags CREATE_AND_SET
+     * @default 0.0.0.0
+     * @validonly SAI_NAT_ENTRY_ATTR_NAT_TYPE == SAI_NAT_TYPE_SOURCE_NAT or SAI_NAT_ENTRY_ATTR_NAT_TYPE == SAI_NAT_TYPE_DOUBLE_NAT
+     */
+    SAI_NAT_ENTRY_ATTR_SRC_IP_MASK,
```

Comment:
Why do we have a mask for setting IP port if the match in the entry is exact?

Reply:
This is for the entire field mask, though ranges can be configured for do not nat use case